src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
+        return false;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public T set(T value, int index) {
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public T get(int index) {
+        checkIndex(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Can't operate on index: " + index
+                    + ", with size: " + size);
+        }
+    }
+
+    private void checkIndexAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Can't operate on index: " + index
+                    + ", with size: " + size);
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        if (index <= size / 2) {
+            Node<T> current = head;
+            int currentIndex = 0;
+            while (current != null && currentIndex < index) {
+                current = current.next;
+                currentIndex++;
+            }
+            return current;
+        } else {
+            Node<T> current = tail;
+            int currentIndex = size - 1;
+            while (current != null && currentIndex > index) {
+                current = current.prev;
+                currentIndex--;
+            }
+            return current;
+        }
+    }
+
+    private void unlink(Node<T> node) {
+        if (node == head) {
+            head = head.next;
+            // якщо головою стало нічого то хвіст також нічого
+            if (head == null) {
+                tail = null;
+            } else {
+                // інакше засетити новій голові нульового сусіда prev
+                head.prev = null;
+            }
+        } else if (node == tail) {
+            // зробити сусіда хвоста хвостом
+            tail = tail.prev;
+            tail.next = null;
+        } else {
+            // сусіду prev призначити next сусіда від next нашої видаленої ноди
+            node.prev.next = node.next;
+            // сусіду next призначити prev сусіда від prev нашої видаленої ноди
+            node.next.prev = node.prev;
```

Comment:
Generaly you dont need to leave a comments in your hw, but if you want to - leave it in english (leaving comments in ukrainian is a bad practice)